modules/pam_motd/pam_motd.c

```diff
@@ -282,6 +283,97 @@ static void try_to_display_directories_with_overrides(pam_handle_t *pamh,
     _pam_drop(dirscans);
 }
 
+static int switch_ids(pam_handle_t *pamh)
+{
+    struct passwd *pw;
+    const char *username;
+    int retval;
+
+    retval = pam_get_user(pamh, &username, "key user");
+
+    if (retval == PAM_SUCCESS) {
+        pw = pam_modutil_getpwnam (pamh, username);
+    } else {
+        goto out;
+    }
+
+    if (setgid(pw->pw_gid) < 0) {
+        pam_syslog(pamh, LOG_ERR, "Unable to change GID to %d\n", pw->pw_gid);
+
+        retval = PAM_SESSION_ERR;
+        goto out;
+    }
+
+    if (setuid(pw->pw_uid) < 0) {
+        pam_syslog(pamh, LOG_ERR, "Unable to change UID to %d\n", pw->pw_uid);
+
+        retval = PAM_SESSION_ERR;
+        goto out;
+    }
+
+out:
+    return retval;
+}
```

Comment:
Please consider using pam_modutil_drop_priv/pam_modutil_regain_priv interface.

Reply:
Thanks! Now the code is neater.